File: Lab 7/Dictionary.java
//Jeffrey Yeung
//jeyyeung
//CMPS 12M
//March 10, 2018
//Binary Search Tree implementation of the Dictionary ADT

public class Dictionary implements DictionaryInterface {
  private class Node {
    String key;
    String value;
    Node left;
    Node right;
    Node(String key, String val) {
      this.key = key;
      this.value = val;
    }
  }

  private Node root;
  private int numItems;
  public Dictionary() {
    root = null;
    numItems = 0;
  }

  public Node findKey(Node R, String k) {
    if(R == null || k.compareTo(R.key) == 0) {
      return R;
    }
    if(k.compareTo(R.key) < 0) {
      return findKey(R.left, k);
    }
    else {
      return findKey(R.right, k);
    }
  }

  public Node findParent(Node N, Node R) {
    Node P = null;
    if(N != R) {
      P = R;
      while(P.left != N && P.right != N) {
        if(N.key.compareToIgnoreCase(P.key) < 0) {
          P = P.left;
        }
        else {
          P = P.right;
        }
      }
    }
    return P;
  }

  public Node findLeftmost(Node R) {
    Node L = R;
    if(L != null) {
      for(int i = 0; L.left != null; i++);
    }
    return L;
  }

  public void printInOrder(Node R) {
    if (R != null) {
      printInOrder(R.left);
      System.out.println(R.key + " " + R.value);
      printInOrder(R.right);
    }
  }

  public void deleteAll() {
    root = null;
    numItems = 0;
  }

  public boolean isEmpty() {
    return (numItems == 0);
  }

  public int size() {
    return numItems;
  }

  public String lookup(String key) {
    Node N = findKey(root, key);
    if(N != null) {
      return N.value;
    }
    else {
      return null;
    }
  }

  public void insert(String key, String value) throws DuplicateKeyException {
    Node N = new Node(key, value);
    Node A = root;
    Node B = null;
    if(findKey(root, key) != null) {
      throw new DuplicateKeyException("Dictionary Error: insert() cannot insert duplicate key: " + key);
    }
    while(A != null) {
      B = A;
      if(key.compareTo(A.key) > 0) {
        A = A.right;
      }
      else {
        A = A.right;
      }
    }
    if(B == null) {
      root = N;
    }
    else if(key.compareTo(B.key) < 0) {
      B.left = N;
    }
    else {
      B.right = N;
    }
    numItems++;
  }

  public void delete(String key) throws KeyNotFoundException {
    Node N = findKey(root, key);
    Node P, S;
    if(findKey(root, key) == null) {
      throw new KeyNotFoundException("Dictionary Error: delete() cannot delete non-existent key: " + key);
    }
    if(N.left == null && N.right == null) {
      if(N == root) {
        root = null;
      }
      else {
        P = findParent(N, root);
        if(P.right == N) {
          P.right = null;
        }
        else {
          P.left = null;
        }
      }
    }
    else if(N.right == null) {
      if(N == root) {
        root = N.left;
      }
      else {
        P = findParent(N, root);
        if(P.right == N) {
          P.right = N.left;
        }
        else {
          P.left = N.left;
        }
      }
    }
    else if(N.left == null) {
      if(N == root) {
        root = N.right;
      }
      else {
        P = findParent(N, root);
        if(P.right == N) {
          P.right = N.right;
        }
        else {
          P.left = N.right;
        }
      }
    }
    else {
      S = findLeftmost(N.right);
      N.key = S.key;
      N.value = S.value;
      P = findParent(S, N);
      if(P.right == S) {
        P.right = S.right;
      }
      else {
        P.left = S.right;
      }
    }
    numItems--;
  }

  public void makeEmpty() {
    deleteAll();
    root = null;
    numItems = 0;
  }

  public String toString() {
    printInOrder(root);
    return "";
  }


}

File: Lab 5/Dictionary.c
//Jeffrey Yeung
//jeyyeung
//CMPS 12M
//February 21, 2018
//implementation for dictionary ADT

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<assert.h>
#include"Dictionary.h"

typedef struct NodeObj{
  char* key;
  char* value;
  struct NodeObj* next;
}
NodeObj;

typedef NodeObj* Node;

Node newNode(char* k, char* v) {
  Node N = malloc(sizeof(NodeObj));
  assert(N!=NULL);
  N->key = k;
  N->value = v;
  N->next = NULL;
  return(N);
}

void freeNode(Node* pN) {
  if( pN!=NULL && *pN!=NULL ) {
    free(*pN);
    *pN = NULL;
  }
}

typedef struct DictionaryObj{
  Node head;
  Node tail;
  int numItems;
}
DictionaryObj;

Dictionary newDictionary(void) {
  Dictionary D = malloc(sizeof(DictionaryObj));
  assert(D!=NULL);
  D->head = NULL;
  D->tail = NULL;
  D->numItems = 0;
  return D;
}

void freeDictionary(Dictionary* pD) {
  if(pD!=NULL && *pD!=NULL ) {
    makeEmpty(*pD);
    free(*pD);
    *pD = NULL;
  }
}

int isEmpty(Dictionary D) {
  if( D == NULL) {
    fprintf(stderr, "Dictionary Error: calling isEmpty() on NULL Dictionary reference\n");
    exit(EXIT_FAILURE);
  }
  return(D->numItems==0);
}

int size(Dictionary D) {
  if( D == NULL) {
    fprintf(stderr, "Dictionary Error: calling size() on NULL Dictionary reference\n");
    exit(EXIT_FAILURE);
  }
  return(D->numItems);
}

char* lookup(Dictionary D, char* k) {
  if( D == NULL) {
    fprintf(stderr, "Dictionary Error: calling lookup() on NULL Dictionary reference\n");
    exit(EXIT_FAILURE);
  }
  Node N = D->head;
  while( N != NULL) {
    if(strcmp(N->key,k)==0){
      return N->value;
    }
    N=N->next;
  }
  return NULL;
}

void insert(Dictionary D, char* k, char* v) {
  if(lookup(D, k) != NULL) {
    fprintf(stderr, "Dictionary Error: calling insert() on NULL Dictionary reference\n");
    exit(EXIT_FAILURE);
  }
  if(D->numItems != 0){
    Node N;
    N = newNode(k, v);
    D->tail->next = N;
    D->tail = N;
  }
  else {
    D->head = D->tail = newNode(k, v);
  }
  D->numItems++;
}

void delete(Dictionary D, char* k) {
  if( lookup(D, k)==NULL) {
    fprintf(stderr, "Dictionary Error: calling lookup() on NULL Dictionary reference\n");
    exit(EXIT_FAILURE);
  }
  D->numItems--;
  if(D->numItems==1) {
    makeEmpty(D);
  }
  else if (strcmp(D->head->key, k)) {
    Node N = D->head;
    Node curr = N;
    while(strcmp(N->key,k)) {
      curr = N;
      N=N->next;
    }
    curr->next=N->next;
    freeNode(&N);
  }
  else {
    Node N = D->head;
    Node curr = N->next;
    freeNode(&N);
    D->head=curr;
  }
}

void makeEmpty(Dictionary D) {
  Node curr1 = D->head;
  Node curr2 = D->head;
  while(curr1 != NULL) {
    curr1 = curr1->next;
    freeNode(&curr2);
    curr2 = curr1;
  }
  D->head = NULL;
  D->tail = NULL;
  D->numItems = 0;
}

void printDictionary(FILE* out, Dictionary D) {
  Node N;
  for(N=D->head; N!=NULL; N=N->next) {
    fprintf(out, "%s %s \n", N->key, N->value);
  }
}
